### PR TITLE
feat: make `create materialized view from source` available

### PR DESCRIPTION
**Pull request by shanicky**
## What's changed and what's your intention?

This PR fixes the `create materialized view from source` feature

It is now possible to `create materialized views` from the rust frontend 

However, the state storage of the source executor is not currently handled and requires a PR from @tabVersion  to complete

Since we need to start the whole MV with the first barrier, we modified `/src/compute/tests/table_v2_materialize.rs` to inject the first barrier

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

### CODE REVIEW

**Comment by neverchanje** [comment]
Do we have e2e tests?

**Comment by shanicky** [comment]
>

**Comment by shanicky** [comment]
> Do we have e2e tests?

e2e will be available in the next pr

File: src/source/src/lib.rs [hunk 1]
```diff
@@ -66,13 +70,64 @@ pub enum SourceImpl {
...
+
+    async fn next(&mut self) -> Result<StreamChunk> {
+        match self {
+            SourceStreamReaderImpl::TableV2(t) => t.next().await,
+            SourceStreamReaderImpl::Connector(c) => c.next().await,
+        }
+    }
+}
+
+pub enum SourceReaderContext {
```
Comment (neverchanje) [inline src/source/src/lib.rs:95]:
```suggestion
pub type SourceReaderContext = Vec<SplitImpl>;
```

None can be represented as empty vec. And what is the difference with `pub(crate) struct ConnectorReaderContext`?

Comment (shanicky) [reply src/source/src/lib.rs:95]: This is a semantic difference; table sources don't need splits to be initialized, so `None` is more appropriate than an empty Vec

File: src/connector/src/kinesis/source/reader.rs [hunk 2]
```diff
@@ -41,7 +41,7 @@ pub struct KinesisSplitReader {
 }
 
 #[async_trait]
-impl SplitReader for KinesisSplitReader {
+impl SourceReader for KinesisSplitReader {
```
Comment (neverchanje) [inline src/connector/src/kinesis/source/reader.rs:44]: SourceReader is ambiguous with the upper-level abstraction StreamSourceReader. SplitReader sounds more low-level and actually it's more aligned with the name of KinesisSplitReader and KafkaSplitReader.

File: src/source/src/connector_source.rs [hunk 3]
```diff
@@ -12,71 +12,118 @@
...
-use crate::{SourceColumnDesc, SourceParserImpl, StreamSourceReader};
+use crate::{Source, SourceColumnDesc, SourceParserImpl, StreamSourceReader};
 
 /// [`ConnectorSource`] serves as a bridge between external components and streaming or batch
-/// processing. [`ConnectorSource`] introduces schema at this level while [`SplitReaderImpl`]
+/// processing. [`ConnectorSource`] introduces schema at this level while [`SourceReaderImpl`]
 /// simply loads raw content from message queue or file system.
 #[derive(Clone)]
 pub struct ConnectorSource {
+    pub config: HashMap<String, String>,
```
Comment (neverchanje) [inline src/source/src/connector_source.rs:34]:
```suggestion
    pub config: Properties,
```

File: src/connector/src/base.rs [hunk 4]
```diff
@@ -62,31 +62,36 @@ pub struct ConnectorState {
...
     pub async fn next(&mut self) -> Result<Option<Vec<SourceMessage>>> {
         match self {
             Self::Kafka(r) => r.next().await,
             Self::Kinesis(r) => r.next().await,
         }
     }
 
-    pub async fn create(config: Properties, state: Option<ConnectorState>) -> Result<Self> {
+    pub async fn create(config: Properties, state: ConnectorStateV2) -> Result<Self> {
+        // pub async fn create(config: Properties, state: Option<ConnectorState>) -> Result<Self> {
```
Comment (neverchanje) [inline src/connector/src/base.rs:94]:
```suggestion
```

File: src/connector/src/base.rs [hunk 5]
```diff
@@ -62,31 +62,36 @@ pub struct ConnectorState {
...
+
 #[async_trait]
-pub trait SplitReader {
+pub trait SourceReader {
     async fn next(&mut self) -> Result<Option<Vec<SourceMessage>>>;
-
-    /// `state` is used to recover from the formerly persisted state.
-    /// If the reader is newly created via CREATE SOURCE, the state will be none.
-    async fn new(properties: Properties, state: Option<ConnectorState>) -> Result<Self>
+    async fn new(properties: Properties, state: ConnectorStateV2) -> Result<Self>
```
Comment (neverchanje) [inline src/connector/src/base.rs:75]: you removed my comments...

File: src/connector/src/base.rs [hunk 6]
```diff
@@ -62,13 +62,19 @@ pub struct ConnectorState {
     pub end_offset: String,
 }
 
+#[derive(Debug, Clone)]
+pub enum ConnectorStateV2 {
```
Comment (neverchanje) [inline src/connector/src/base.rs:66]: This is too dirty. We definitely need another implementation or another name later.

**Pull request merged**